File: server/src/main/java/yangFenTuoZi/runner/plus/server/FakeContext.java
package yangFenTuoZi.runner.plus.server;

import android.annotation.TargetApi;
import androidx.annotation.NonNull;

import android.content.AttributionSource;
import android.content.Context;
import android.content.ContextWrapper;
import android.os.Build;
import android.os.Process;
import android.system.Os;

public final class FakeContext extends ContextWrapper {
    public static String PACKAGE_NAME = Os.getuid() == 0 ? "root" : "com.android.shell";
    private static final FakeContext INSTANCE = new FakeContext();

    public static FakeContext get() {
        return INSTANCE;
    }

    private FakeContext() {
        super(Workarounds.getSystemContext());
    }

    @Override
    public String getPackageName() {
        return PACKAGE_NAME;
    }

    @Override
    @NonNull
    public String getOpPackageName() {
        return PACKAGE_NAME;
    }

    @TargetApi(Build.VERSION_CODES.S)
    @Override
    @NonNull
    public AttributionSource getAttributionSource() {
        AttributionSource.Builder builder = new AttributionSource.Builder(Process.SHELL_UID);
        builder.setPackageName(PACKAGE_NAME);
        return builder.build();
    }

    // @Override to be added on SDK upgrade for Android 14
    @SuppressWarnings("unused")
    public int getDeviceId() {
        return 0;
    }

    @Override
    public Context getApplicationContext() {
        return this;
    }
}

File: app/src/main/java/yangfentuozi/runner/server/fakecontext/FakeContext.kt
package yangfentuozi.runner.server.fakecontext

import android.content.AttributionSource
import android.content.Context
import android.content.ContextWrapper
import android.os.Build
import android.system.Os
import androidx.annotation.RequiresApi

class FakeContext private constructor() : ContextWrapper(Workarounds.systemContext) {
    override fun getPackageName(): String {
        return PACKAGE_NAME
    }

    override fun getOpPackageName(): String {
        return PACKAGE_NAME
    }

    @RequiresApi(Build.VERSION_CODES.S)
    override fun getAttributionSource(): AttributionSource {
        val builder = AttributionSource.Builder(Os.getuid())
        builder.setPackageName(PACKAGE_NAME)
        return builder.build()
    }

    // @Override to be added on SDK upgrade for Android 14
    @Suppress("unused")
    override fun getDeviceId(): Int {
        return 0
    }

    override fun getApplicationContext(): Context {
        return this
    }

    companion object {
        var PACKAGE_NAME: String = if (Os.getuid() == 0) "root" else "com.android.shell"
        private val INSTANCE = FakeContext()

        fun get(): FakeContext {
            return INSTANCE
        }
    }
}